### PR TITLE
minor fix for stablility

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/agentchat/conversable_agent.py [hunk 1]
```diff
@@ -606,6 +606,8 @@ def generate_code_execution_reply(
     ):
         """Generate a reply using code execution."""
         code_execution_config = config if config is not None else self._code_execution_config
+        exitcode = None
```
Comment (LittleLittleCloud) [inline autogen/agentchat/conversable_agent.py:609]: create placeholder for exitcode and exitcode2str so it won't crash when there's 1 unknown code block

File: autogen/code_utils.py [hunk 2]
```diff
@@ -252,7 +259,7 @@ def execute_code(
     file_dir = os.path.dirname(filepath)
     os.makedirs(file_dir, exist_ok=True)
     if code is not None:
-        with open(filepath, "w") as fout:
+        with open(filepath, "w", encoding='utf-8') as fout:
```
Comment (LittleLittleCloud) [inline autogen/code_utils.py:262]: will crash on windows if code contains non-ascill char without explicilty setting encoding

File: autogen/agentchat/conversable_agent.py (hunk 1 above)
Comment (sonichi) [reply autogen/agentchat/conversable_agent.py:609]: The root cause is in line 622. The condition should be checking whether i is the last in the range. Please fix line 622 and remove the change here.

File: autogen/agentchat/conversable_agent.py [hunk 3]
```diff
@@ -611,25 +611,23 @@ def generate_code_execution_reply(
...
+        # if code blocks are found, execute the code blocks and return the output
+        # if no code blocks are found, continue
         for i in range(min(len(messages), last_n_messages)):
             message = messages[-(i + 1)]
             code_blocks = extract_code(message["content"])
             if len(code_blocks) == 1 and code_blocks[0][0] == UNKNOWN:
-                # no code block is found, lang should be `UNKNOWN`
-
-                if i == last_n_messages - 1:
-                    code_execution_config["last_n_messages"] = last_n_messages
```
Comment (sonichi) [inline autogen/agentchat/conversable_agent.py:621]: this line shouldn't be removed.

File: autogen/agentchat/conversable_agent.py [hunk 4]
```diff
@@ -611,25 +611,23 @@ def generate_code_execution_reply(
...
-                # code_blocks, _ = find_code(messages, sys_msg=self._oai_system_message, **self.llm_config)
-                # if len(code_blocks) == 1 and code_blocks[0][0] == UNKNOWN:
-                #     return code_blocks[0][1]
-            # try to execute the code
+
+            # found code blocks, try to execute the code
             exitcode, logs = self.execute_code_blocks(code_blocks)
             exitcode2str = "execution succeeded" if exitcode == 0 else "execution failed"
-            break
-        code_execution_config["last_n_messages"] = last_n_messages
```
Comment (sonichi) [inline autogen/agentchat/conversable_agent.py:631]: this line before return shouldn't be removed.

File: autogen/agentchat/conversable_agent.py (hunk 3 above)
Comment (LittleLittleCloud) [reply autogen/agentchat/conversable_agent.py:621]: What's the logic for searching code block here

I assume it's searching the 'last_n_messages' from `code_execution_config` reversely and run the first code block that detected? In that case `last_n_messages` would be a read-only config and shouldn't need to be written back?

Comment (sonichi) [reply autogen/agentchat/conversable_agent.py:621]: The variable is popped out of the dict before the loop starts. So we need to restore it before returning.

**Comment by sonichi** [comment]
To run the OpenAI test, we need the PR to be made from a branch in the original repo. Could you make the PR that way? Or if you know an alternative that allows running the test even from forked repo, please suggest.

**Comment by LittleLittleCloud** [comment]
> To run the OpenAI test, we need the PR to be made from a branch in the original repo. Could you make the PR that way? Or if you know an alternative that allows running the test even from forked repo, please suggest.

To use the context from the base of repo to fork repo when running an action or workflow, you can use [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) as trigger. Make sure you add safety check (like additional approval before running OpenAI job) because running action from untrusted source might cause secret leak. You can find more information from [this blog](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)

**Comment by LittleLittleCloud** [comment]
> > > To run the OpenAI test, we need the PR to be made from a branch in the original repo. Could you make the PR that way? Or if you know an alternative that allows running the test even from forked repo, please suggest.
> > 
> > 
> > To use the context from the base of repo to fork repo when running an action or workflow, you can use [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) as trigger. Make sure you add safety check (like additional approval before running OpenAI job) because running action from untrusted source might cause secret leak. You can find more information from [this blog](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)
> 
> I've tried this before but it didn't work. Feel free to give it another try if you want.

I can take a look, I'll need permission to push to original repo though, can you grant me that (contributor or admin should both work)

**Comment by sonichi** [comment]
> > > > To run the OpenAI test, we need the PR to be made from a branch in the original repo. Could you make the PR that way? Or if you know an alternative that allows running the test even from forked repo, please suggest.
> > > 
> > > 
> > > To use the context from the base of repo to fork repo when running an action or workflow, you can use [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) as trigger. Make sure you add safety check (like additional approval before running OpenAI job) because running action from untrusted source might cause secret leak. You can find more information from [this blog](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)
> > 
> > 
> > I've tried this before but it didn't work. Feel free to give it another try if you want.
> 
> I can take a look, I'll need permission to push to original repo though, can you grant me that (contributor or admin should both work)

you should have maintainer permission already.

**Comment by LittleLittleCloud** [comment]
> > > > > To run the OpenAI test, we need the PR to be made from a branch in the original repo. Could you make the PR that way? Or if you know an alternative that allows running the test even from forked repo, please suggest.
> > > > 
> > > > 
> > > > To use the context from the base of repo to fork repo when running an action or workflow, you can use [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) as trigger. Make sure you add safety check (like additional approval before running OpenAI job) because running action from untrusted source might cause secret leak. You can find more information from [this blog](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)
> > > 
> > > 
> > > I've tried this before but it didn't work. Feel free to give it another try if you want.
> > 
> > 
> > I can take a look, I'll need permission to push to original repo though, can you grant me that (contributor or admin should both work)
> 
> you should have maintainer permission already.

Did you enable `Required reviewers` on OpenAI environemnt? Can't see that environment page on my end
image

**Comment by sonichi** [comment]
> > > > > > To run the OpenAI test, we need the PR to be made from a branch in the original repo. Could you make the PR that way? Or if you know an alternative that allows running the test even from forked repo, please suggest.
> > > > > 
> > > > > 
> > > > > To use the context from the base of repo to fork repo when running an action or workflow, you can use [`pull_request_target`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) as trigger. Make sure you add safety check (like additional approval before running OpenAI job) because running action from untrusted source might cause secret leak. You can find more information from [this blog](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)
> > > > 
> > > > 
> > > > I've tried this before but it didn't work. Feel free to give it another try if you want.
> > > 
> > > 
> > > I can take a look, I'll need permission to push to original repo though, can you grant me that (contributor or admin should both work)
> > 
> > 
> > you should have maintainer permission already.
> 
> Did you enable `Required reviewers` on OpenAI environemnt? Can't see that environment page on my end image

Yes I did.

**Comment by LittleLittleCloud** [comment]
Cool, then you can go check #21 out. The secureness of secrets should be able to be protected by `Environemnt Protection Rule`.

Note that this still brings side effects, as the external source will still be run on target environment once it gets approval, so the secureness will rely on reviewers, who might make mistakes. It's the cost we have to pay if we want a contribution from the forked repo though.

In order to add another layer of secure, you might want to enable `Require approval from all outside collaborator` as well
image

**Comment by LittleLittleCloud** [comment]
I'm going to create another PR from main repo to avoid blocking by #21

**Pull request closed without merging**